pkg/emitter/nats_emitter_test.go

```diff
@@ -0,0 +1,197 @@
+//
+// Copyright 2022 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package emitter
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/guacsec/guac/internal/testing/dochelper"
+	nats_test "github.com/guacsec/guac/internal/testing/nats"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/logging"
+	uuid "github.com/satori/go.uuid"
+)
+
+var (
+	// Taken from: https://slsa.dev/provenance/v0.1#example
+	ite6SLSA = `
+	{
+		"_type": "https://in-toto.io/Statement/v0.1",
+		"subject": [{"name": "helloworld", "digest": {"sha256": "5678..."}}],
+		"predicateType": "https://slsa.dev/provenance/v0.2",
+		"predicate": {
+			"builder": { "id": "https://github.com/Attestations/GitHubHostedActions@v1" },
+			"buildType": "https://github.com/Attestations/GitHubActionsWorkflow@v1",
+			"invocation": {
+			  "configSource": {
+				"uri": "git+https://github.com/curl/curl-docker@master",
+				"digest": { "sha1": "d6525c840a62b398424a78d792f457477135d0cf" },   
+				"entryPoint": "build.yaml:maketgz"
+			  }
+			},
+			"metadata": {
+			  "buildStartedOn": "2020-08-19T08:38:00Z",
+			  "completeness": {
+				  "environment": true
+			  }
+			},
+			"materials": [
+			  {
+				"uri": "git+https://github.com/curl/curl-docker@master",
+				"digest": { "sha1": "d6525c840a62b398424a78d792f457477135d0cf" }
+			  }, {
+				"uri": "github_hosted_vm:ubuntu-18.04:20210123.1",
+				"digest": { "sha1": "d6525c840a62b398424a78d792f457477135d0cf" }
+			  }
+			]
+		}
+	}`
+
+	ite6SLSADoc = processor.Document{
+		Blob:   []byte(ite6SLSA),
+		Type:   processor.DocumentITE6SLSA,
+		Format: processor.FormatJSON,
+		SourceInformation: processor.SourceInformation{
+			Collector: "TestCollector",
+			Source:    "TestSource",
+		},
+	}
+)
+
+func TestNatsEmitter_PublishOnEmit(t *testing.T) {
+	expectedDocTree := dochelper.DocNode(&ite6SLSADoc)
+
+	natsTest := nats_test.NewNatsTestServer()
+	url, err := natsTest.EnableJetStreamForTest()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer natsTest.Shutdown()
+
+	ctx := context.Background()
+	jetStream := NewJetStream(url, "", "")
+	ctx, err = jetStream.JetStreamInit(ctx)
+	if err != nil {
+		t.Fatalf("unexpected error initializing jetstream: %v", err)
+	}
+	err = jetStream.RecreateStream(ctx)
+	if err != nil {
+		t.Fatalf("unexpected error recreating jetstream: %v", err)
+	}
```

Comment:
nit: probably better as its own test, since it's weird sticking in the context of the pub/sub test... unless i'm missing something.

Reply:
This is used in the test to ensure any lingering documents do get added to the test. I did create another separate test just for `RecreateStream` also.